File: src/integrationTest/java/net/ltgt/gradle/nullaway/BinaryCompatibilityIntegrationTest.java
package net.ltgt.gradle.nullaway;

import static com.google.common.truth.Truth.assertThat;
import static java.util.Objects.requireNonNull;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Stream;
import org.gradle.testkit.runner.BuildResult;
import org.gradle.testkit.runner.GradleRunner;
import org.gradle.testkit.runner.TaskOutcome;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

public class BinaryCompatibilityIntegrationTest extends BaseIntegrationTest {
  @TempDir Path conventionPluginProjectDir;
  Path conventionPluginBuildFile;

  @BeforeEach
  void setup() throws Exception {
    Files.writeString(
        getBuildFile(),
        // language=kts
        """
        plugins {
            `java-library`
            id("net.ltgt.errorprone")
            id("net.ltgt.nullaway")
            id("convention.nullaway")
        }

        repositories {
            mavenCentral()
        }

        dependencies {
            errorprone("com.google.errorprone:error_prone_core:%s")
            errorprone("com.uber.nullaway:nullaway:%s")
        }

        tasks.withType<JavaCompile>().configureEach {
            options.compilerArgs.add("-Werror")
        }

        if (GradleVersion.current() < GradleVersion.version("7.0")) {
            allprojects {
                configurations.all {
                    attributes.attribute(Attribute.of("org.gradle.jvm.environment", String::class.java), "standard-jvm")
                }
            }
        }
        """
            .formatted(errorproneVersion, nullawayVersion));
    Files.copy(
        projectDir.resolve("gradle.properties"),
        conventionPluginProjectDir.resolve("gradle.properties"));
    Files.writeString(
        conventionPluginProjectDir.resolve("settings.gradle.kts"),
        // language=kts
        """
        rootProject.name = "convention-plugin"
        """);
    conventionPluginBuildFile =
        Files.createFile(conventionPluginProjectDir.resolve("build.gradle.kts"));
  }

  private void buildConventionPlugin() {
    GradleRunner.create()
        .withGradleVersion(testGradleVersion.getVersion())
        .withProjectDir(conventionPluginProjectDir.toFile())
        .withArguments("assemble")
        .build();
  }

  private BuildResult buildProject(String... args) {
    var runner =
        GradleRunner.create()
            .withGradleVersion(testGradleVersion.getVersion())
            .withProjectDir(projectDir.toFile())
            .withPluginClasspath();
    return runner
        .withPluginClasspath(
            Stream.concat(
                    runner.getPluginClasspath().stream(),
                    Stream.of(
                        conventionPluginProjectDir
                            .resolve("build/libs/convention-plugin.jar")
                            .toFile()))
                .toList())
        .withArguments(args)
        .build();
  }

  @Test
  void isBinaryCompatibleWithPreviousVersions_kotlinDslFlavor() throws Exception {
    // given
    Files.writeString(
        conventionPluginBuildFile,
        // language=kts
        """
        plugins {
            `kotlin-dsl`
        }

        repositories {
            gradlePluginPortal()
        }

        dependencies {
            implementation("net.ltgt.errorprone:net.ltgt.errorprone.gradle.plugin:4.4.0")
            implementation("net.ltgt.nullaway:net.ltgt.nullaway.gradle.plugin:2.4.1")
        }
        """);
    Files.writeString(
        Files.createDirectories(conventionPluginProjectDir.resolve("src/main/kotlin"))
            .resolve("convention.nullaway.gradle.kts"),
        // language=kts
        """
        import net.ltgt.gradle.errorprone.*
        import net.ltgt.gradle.nullaway.*

        plugins {
            id("net.ltgt.errorprone")
            id("net.ltgt.nullaway")
        }

        configure<NullAwayExtension> {
            annotatedPackages.add("test")
        }

        tasks.withType<JavaCompile>().configureEach {
            options.errorprone.nullaway {
                severity.set(CheckSeverity.DEFAULT)
                onlyNullMarked.set(false)
                unannotatedSubPackages.add("test.dummy")
                unannotatedClasses.add("test.Unannotated")
                knownInitializers.add("com.foo.Bar.method")
                excludedClassAnnotations.add("com.example.NullAwayExcluded")
                excludedClasses.add("test.Excluded")
                excludedFieldAnnotations.add("javax.ws.rs.core.Context")
                customInitializerAnnotations.add("com.foo.Initializer")
                externalInitAnnotations.add("com.example.ExternalInit")
                treatGeneratedAsUnannotated.set(true)
                acknowledgeRestrictiveAnnotations.set(true)
                checkOptionalEmptiness.set(true)
                suggestSuppressions.set(true)
                isAssertsEnabled.set(true)
                isExhaustiveOverride.set(true)
                castToNonNullMethod.set("com.foo.Bar.castToNonNull")
                checkOptionalEmptinessCustomClasses.add("com.foo.Optional")
                autoFixSuppressionComment.set("Auto-fix\\\\u0020suppression")
                handleTestAssertionLibraries.set(true)
                acknowledgeAndroidRecent.set(true)
                checkContracts.set(true)
                customContractAnnotations.add("com.example.Contract")
                customNullableAnnotations.add("com.example.CouldBeNull")
                customNonnullAnnotations.add("com.example.MustNotBeNull")
                isJSpecifyMode.set(true)
                extraFuturesClasses.add("com.example.Future")
                suppressionNameAliases.add("NullIssue")
                warnOnGenericInferenceFailure.set(true)
            }
        }
        """);

    buildConventionPlugin();

    writeSuccessSource();

    // when
    var result = buildProject("compileJava");

    // then
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.SUCCESS);
  }

  @Test
  void isBinaryCompatibleWithPreviousVersions_javaFlavor() throws Exception {
    // given
    Files.writeString(
        conventionPluginBuildFile,
        // language=kts
        """
        plugins {
            `java-gradle-plugin`
        }

        repositories {
            gradlePluginPortal()
        }

        dependencies {
            implementation("net.ltgt.errorprone:net.ltgt.errorprone.gradle.plugin:4.4.0")
            implementation("net.ltgt.nullaway:net.ltgt.nullaway.gradle.plugin:2.4.1")
        }

        gradlePlugin {
            plugins {
                create("conventionPlugin") {
                    id = "convention.nullaway"
                    implementationClass = "convention.ConventionNullAwayPlugin"
                }
            }
        }
        """);
    Files.writeString(
        Files.createDirectories(conventionPluginProjectDir.resolve("src/main/java/convention"))
            .resolve("ConventionNullAwayPlugin.java"),
        // language=java
        """
        package convention;

        import net.ltgt.gradle.errorprone.*;
        import net.ltgt.gradle.nullaway.*;
        import org.gradle.api.Plugin;
        import org.gradle.api.Project;
        import org.gradle.api.plugins.ExtensionAware;
        import org.gradle.api.tasks.compile.JavaCompile;

        public class ConventionNullAwayPlugin implements Plugin<Project> {
          @Override
          public void apply(Project project) {
            project.getPluginManager().apply("net.ltgt.errorprone");
            project.getPluginManager().apply("net.ltgt.nullaway");

            project
                .getExtensions()
                .configure(
                    NullAwayExtension.class,
                    nullaway -> {
                      nullaway.getAnnotatedPackages().add("test");
                    });

            project
                .getTasks()
                .withType(JavaCompile.class)
                .configureEach(
                    task -> {
                      ErrorProneOptions errorproneOptions =
                          ((ExtensionAware) task.getOptions())
                              .getExtensions()
                              .getByType(ErrorProneOptions.class);
                      NullAwayOptions nullawayOptions =
                          ((ExtensionAware) errorproneOptions)
                              .getExtensions()
                              .getByType(NullAwayOptions.class);
                      nullawayOptions.getSeverity().set(CheckSeverity.DEFAULT);
                      nullawayOptions.getOnlyNullMarked().set(false);
                      nullawayOptions.getUnannotatedSubPackages().add("test.dummy");
                      nullawayOptions.getUnannotatedClasses().add("test.Unannotated");
                      nullawayOptions.getKnownInitializers().add("com.foo.Bar.method");
                      nullawayOptions.getExcludedClassAnnotations().add("com.example.NullAwayExcluded");
                      nullawayOptions.getExcludedClasses().add("test.Excluded");
                      nullawayOptions.getExcludedFieldAnnotations().add("javax.ws.rs.core.Context");
                      nullawayOptions.getCustomInitializerAnnotations().add("com.foo.Initializer");
                      nullawayOptions.getExternalInitAnnotations().add("com.example.ExternalInit");
                      nullawayOptions.getTreatGeneratedAsUnannotated().set(true);
                      nullawayOptions.getAcknowledgeRestrictiveAnnotations().set(true);
                      nullawayOptions.getCheckOptionalEmptiness().set(true);
                      nullawayOptions.getSuggestSuppressions().set(true);
                      nullawayOptions.getAssertsEnabled().set(true);
                      nullawayOptions.getExhaustiveOverride().set(true);
                      nullawayOptions.getCastToNonNullMethod().set("com.foo.Bar.castToNonNull");
                      nullawayOptions.getCheckOptionalEmptinessCustomClasses().add("com.foo.Optional");
                      nullawayOptions.getAutoFixSuppressionComment().set("Auto-fix\\\\u0020suppression");
                      nullawayOptions.getHandleTestAssertionLibraries().set(true);
                      nullawayOptions.getAcknowledgeAndroidRecent().set(true);
                      nullawayOptions.getCheckContracts().set(true);
                      nullawayOptions.getCustomContractAnnotations().add("com.example.Contract");
                      nullawayOptions.getCustomNullableAnnotations().add("com.example.CouldBeNull");
                      nullawayOptions.getCustomNonnullAnnotations().add("com.example.MustNotBeNull");
                      nullawayOptions.getJspecifyMode().set(true);
                      nullawayOptions.getExtraFuturesClasses().add("com.example.Future");
                      nullawayOptions.getSuppressionNameAliases().add("NullIssue");
                      nullawayOptions.getWarnOnGenericInferenceFailure().set(true);
                    });
          }
        }
        """);

    buildConventionPlugin();

    writeSuccessSource();

    // when
    var result = buildProject("compileJava");

    // then
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.SUCCESS);
  }
}

File: src/integrationTest/kotlin/net/ltgt/gradle/nullaway/BinaryCompatibilityIntegrationTest.kt
package net.ltgt.gradle.nullaway

import com.google.common.truth.Truth.assertThat
import org.gradle.testkit.runner.GradleRunner
import org.gradle.testkit.runner.TaskOutcome
import org.gradle.util.GradleVersion
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.util.Properties

class BinaryCompatibilityIntegrationTest {
    @TempDir
    lateinit var testProjectDir: File
    lateinit var settingsFile: File
    lateinit var buildFile: File

    @TempDir
    lateinit var conventionPluginProjectDir: File
    lateinit var conventionPluginBuildFile: File

    @BeforeEach
    fun setupProject() {
        assumeCompatibleGradleAndJavaVersions()
        testProjectDir.resolve("gradle.properties").outputStream().use {
            Properties().apply {
                setProperty("org.gradle.java.home", testJavaHome)
                store(it, null)
            }
        }
        settingsFile =
            testProjectDir.resolve("settings.gradle.kts").apply {
                createNewFile()
            }
        buildFile =
            testProjectDir.resolve("build.gradle.kts").apply {
                writeText(
                    """
                    plugins {
                        `java-library`
                        id("net.ltgt.errorprone")
                        id("net.ltgt.nullaway")
                        id("convention.nullaway")
                    }

                    repositories {
                        mavenCentral()
                    }

                    dependencies {
                        errorprone("com.google.errorprone:error_prone_core:$errorproneVersion")
                        errorprone("com.uber.nullaway:nullaway:$nullawayVersion")
                    }

                    tasks.withType<JavaCompile>().configureEach {
                        options.compilerArgs.add("-Werror")
                    }
                    """.trimIndent(),
                )
            }
        if (testGradleVersion < GradleVersion.version("7.0")) {
            buildFile.appendText(
                """

                allprojects {
                    configurations.all {
                        attributes.attribute(Attribute.of("org.gradle.jvm.environment", String::class.java), "standard-jvm")
                    }
                }
                """.trimIndent(),
            )
        }
        testProjectDir.resolve("gradle.properties").copyTo(conventionPluginProjectDir.resolve("gradle.properties"))
        conventionPluginProjectDir.resolve("settings.gradle.kts").apply {
            writeText(
                """
                rootProject.name = "convention-plugin"
                """.trimIndent(),
            )
        }
        conventionPluginBuildFile =
            conventionPluginProjectDir.resolve("build.gradle.kts").apply {
                createNewFile()
            }
    }

    private fun buildConventionPlugin() {
        GradleRunner
            .create()
            .withGradleVersion(testGradleVersion.version)
            .withProjectDir(conventionPluginProjectDir)
            .withArguments("assemble")
            .build()
    }

    private fun buildProject(vararg tasks: String) =
        GradleRunner
            .create()
            .withGradleVersion(testGradleVersion.version)
            .withProjectDir(testProjectDir)
            .withPluginClasspath()
            .apply {
                withPluginClasspath(pluginClasspath + conventionPluginProjectDir.resolve("build/libs/convention-plugin.jar"))
            }.withArguments(*tasks)
            .build()

    @Test
    fun `is binary-compatible with previous versions (kotlin-dsl flavor)`() {
        // given
        conventionPluginBuildFile.writeText(
            """
            plugins {
                `kotlin-dsl`
            }

            repositories {
                gradlePluginPortal()
            }

            dependencies {
                implementation("net.ltgt.errorprone:net.ltgt.errorprone.gradle.plugin:4.4.0")
                implementation("net.ltgt.nullaway:net.ltgt.nullaway.gradle.plugin:2.4.1")
            }
            """.trimIndent(),
        )
        conventionPluginProjectDir.resolve("src/main/kotlin").apply { mkdirs() }.resolve("convention.nullaway.gradle.kts").apply {
            writeText(
                """
                import net.ltgt.gradle.errorprone.*
                import net.ltgt.gradle.nullaway.*

                plugins {
                    id("net.ltgt.errorprone")
                    id("net.ltgt.nullaway")
                }

                configure<NullAwayExtension> {
                    annotatedPackages.add("test")
                }

                tasks.withType<JavaCompile>().configureEach {
                    options.errorprone.nullaway {
                        isAssertsEnabled.set(true)
                        isExhaustiveOverride.set(true)
                        isJSpecifyMode.set(true)
                    }
                }
                """.trimIndent(),
            )
        }

        buildConventionPlugin()

        testProjectDir.writeSuccessSource()

        // when
        val result = buildProject(":compileJava", "--stacktrace")

        // then
        assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
    }

    @Test
    fun `is binary-compatible with previous versions (java flavor)`() {
        // given
        conventionPluginBuildFile.writeText(
            """
            plugins {
                `java-gradle-plugin`
            }

            repositories {
                gradlePluginPortal()
            }

            dependencies {
                implementation("net.ltgt.errorprone:net.ltgt.errorprone.gradle.plugin:4.4.0")
                implementation("net.ltgt.nullaway:net.ltgt.nullaway.gradle.plugin:2.4.1")
            }
            
            gradlePlugin {
                plugins {
                    create("conventionPlugin") {
                        id = "convention.nullaway"
                        implementationClass = "convention.ConventionNullAwayPlugin"
                    }
                }
            }
            """.trimIndent(),
        )
        conventionPluginProjectDir.resolve("src/main/java/convention").apply { mkdirs() }.resolve("ConventionNullAwayPlugin.java").apply {
            writeText(
                """
                package convention;

                import net.ltgt.gradle.errorprone.*;
                import net.ltgt.gradle.nullaway.*;
                import org.gradle.api.Plugin;
                import org.gradle.api.Project;
                import org.gradle.api.plugins.ExtensionAware;
                import org.gradle.api.tasks.compile.JavaCompile;
                
                public class ConventionNullAwayPlugin implements Plugin<Project> {
                  @Override
                  public void apply(Project project) {
                    project.getPluginManager().apply("net.ltgt.errorprone");
                    project.getPluginManager().apply("net.ltgt.nullaway");
                
                    project
                        .getExtensions()
                        .configure(
                            NullAwayExtension.class,
                            nullaway -> {
                              nullaway.getAnnotatedPackages().add("test");
                            });
                
                    project
                        .getTasks()
                        .withType(JavaCompile.class)
                        .configureEach(
                            task -> {
                              ErrorProneOptions errorproneOptions =
                                  ((ExtensionAware) task.getOptions())
                                      .getExtensions()
                                      .getByType(ErrorProneOptions.class);
                              NullAwayOptions nullawayOptions =
                                  ((ExtensionAware) errorproneOptions)
                                      .getExtensions()
                                      .getByType(NullAwayOptions.class);
                              nullawayOptions.getAssertsEnabled().set(true);
                              nullawayOptions.getExhaustiveOverride().set(true);
                              nullawayOptions.getJspecifyMode().set(true);
                            });
                  }
                }
                """.trimIndent(),
            )
        }

        buildConventionPlugin()

        testProjectDir.writeSuccessSource()

        // when
        val result = buildProject(":compileJava", "--stacktrace")

        // then
        assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
    }
}
